include/ginkgo/core/log/batch_logger.hpp

```diff
@@ -33,39 +33,12 @@ namespace detail {
 template <typename ValueType>
 struct log_data final {
     using real_type = remove_complex<ValueType>;
+    using index_type = int;
 
-    log_data(std::shared_ptr<const Executor> exec, size_type num_batch_items)
-        : res_norms(exec), iter_counts(exec)
-    {
-        const size_type workspace_size =
-            num_batch_items * (sizeof(real_type) + sizeof(int));
-        if (num_batch_items > 0) {
-            iter_counts.resize_and_reset(num_batch_items);
-            res_norms.resize_and_reset(num_batch_items);
-        } else {
-            GKO_INVALID_STATE("Invalid num batch items passed in");
-        }
-    }
+    log_data(std::shared_ptr<const Executor> exec, size_type num_batch_items);
 
     log_data(std::shared_ptr<const Executor> exec, size_type num_batch_items,
-             array<unsigned char>& workspace)
-        : res_norms(exec), iter_counts(exec)
-    {
-        const size_type workspace_size =
-            num_batch_items * (sizeof(real_type) + sizeof(int));
-        if (num_batch_items > 0 && !workspace.is_owning() &&
-            workspace.get_size() >= workspace_size) {
-            iter_counts =
-                array<int>::view(exec, num_batch_items,
-                                 reinterpret_cast<int*>(workspace.get_data()));
-            res_norms = array<real_type>::view(
-                exec, num_batch_items,
-                reinterpret_cast<real_type*>(workspace.get_data() +
-                                             (sizeof(int) * num_batch_items)));
```

Comment:
originally, I thought it should be solved by using the view on the larger type first.

Reply:
I havent tested that, but it is possible, but for cases when we need to have different types, it would not help. This approach is more general.

Reply:
one `if-else` condition should solve all cases

Reply:
The plan would be to use this in all cases, for shared memory inside the kernels as well, so IMO CUB solution is a better and more general approach.